internal/pkg/agent/cmd/container_init_linux.go

```diff
@@ -0,0 +1,283 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/elastic/elastic-agent/pkg/utils"
+
+	"kernel.org/pub/linux/libs/security/libcap/cap"
+
+	"github.com/elastic/elastic-agent/internal/pkg/cli"
+)
+
+type capProc interface {
+	GetFlag(vec cap.Flag, val cap.Value) (bool, error)
+	SetFlag(vec cap.Flag, enable bool, val ...cap.Value) error
+	SetProc() error
+}
+
+type capBound interface {
+	SetVector(vec cap.Vector, raised bool, vals ...cap.Value) error
+	SetProc() error
+}
+
+var (
+	// for unit-testing
+	capBoundFunc = func() capBound {
+		return cap.NewIAB()
+	}
+	// for unit-testing
+	capProcFunc = func() capProc {
+		return cap.GetProc()
+	}
+)
+
+func logWarning(streams *cli.IOStreams, err error) {
+	fmt.Fprintf(streams.Err, "Warning: %v\n", err)
+}
+
+// initContainer applies the following container initialisation steps:
+//   - raises the capabilities of the Effective and Inheritable sets to match the ones in the Permitted set
+//   - raises the capabilities of the Ambient set to match the ones in Effective set
+//   - chown all agent-related paths
+//
+// Note that to avoid disrupting effects, any error is logged as a warning, but not returned.
+func initContainer(streams *cli.IOStreams) {
+	isRoot, err := utils.HasRoot()
+	if err != nil {
+		logWarning(streams, err)
+		return
+	}
+	if !isRoot {
+		logInfo(streams, "agent container initialisation - effective capabilities")
+		if err := raiseEffectiveCapabilities(); err != nil {
+			logWarning(streams, err)
+			return
+		}
+
+		logInfo(streams, "agent container initialisation - ambient capabilities")
+		if err := raiseAmbientCapabilities(); err != nil {
+			logWarning(streams, err)
+		}
+	}
+
+	// we need to chown all paths
+	logInfo(streams, "agent container initialisation - chown paths")
+	if err := chownPaths(agentBaseDirectory); err != nil {
+		logWarning(streams, err)
```

Comment:
Is this one actually fatal, or can we detect situations where it is likely to lead to problems? Like mounted files being owned by root and agent not being root?

Reply:
I was thinking about the same, but I end up in the conclusion that this doesn't need to become fatal here because it is gonna become fatal later on when agent tries to touch/write something that it can't due to permissions 😄 thus I resorted to a warning like "hey grant this permission if you face any ownership problems"

Reply:
Yeah my concern is that this error is very specific and includes the correction a user needs to make, and whatever we get later is going to be way more vague and this log may have rotated away or be easy to miss.

Reply:
the thing that I am not entirely sure how to handle is that to detect imminent access problems I have to know if a given file will be read, written or both from the agent/agentbeat. This is somewhat tricky to capture and if you add also the access needs of individual beats through agentbeat becomes more tricky?! So not plenty of ideas at the moment 🙂